coax/reward_tracing/_nstep_verbose.py

```diff
@@ -0,0 +1,73 @@
+# MIT License                                                                                      #
+#                                                                                                  #
+# Copyright (c) 2020, Microsoft Corporation                                                        #
+#                                                                                                  #
+# Permission is hereby granted, free of charge, to any person obtaining a copy of this software    #
+# and associated documentation files (the "Software"), to deal in the Software without             #
+# restriction, including without limitation the rights to use, copy, modify, merge, publish,       #
+# distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the    #
+# Software is furnished to do so, subject to the following conditions:                             #
+#                                                                                                  #
+# The above copyright notice and this permission notice shall be included in all copies or         #
+# substantial portions of the Software.                                                            #
+#                                                                                                  #
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING    #
+# BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND       #
+# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,     #
+# DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,   #
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.          #
+# ------------------------------------------------------------------------------------------------ #
+
+from itertools import islice
+import jax.numpy as jnp
+
+from .._base.errors import InsufficientCacheError
+from ._nstep import NStep
+from ._transition import TransitionBatch
+
+__all__ = (
+    'NStep',
+)
+
+
+class NStepVerbose(NStep):
+
+    def pop(self):
+        if not self:
+            raise InsufficientCacheError(
+                "cache needs to receive more transitions before it can be popped from")
+
+        # pop state-action (propensities) pair
+        s, a, logp, w = self._deque_s.popleft()
+
+        # n-step partial return
+        zipped = zip(self._gammas, self._deque_r)
+        rn = sum(x * r for x, r in islice(zipped, self.n))
+        self._deque_r.popleft()
+
+        states = [s]
+        dones = [False]
+        i = 0
+        while i < self.n and i < len(self):
+            s_next, *_ = self._deque_s[i]
+            states.append(s_next)
+            dones.append(False)
+            i += 1
+        while len(states) != self.n + 1:
+            states.append(s)
```

Comment:
I just noticed that this leads to the entropy of the t+n'th state being added to the regularization which is false. I will change this.

Reply:
Could you use the `dones` to mask those contributions?

Reply:
Yes, I added some code to filter those!